Source/scrollrt.cpp

```diff
@@ -946,38 +946,57 @@ static void scrollrt_draw(CelOutputBuffer out, int x, int y, int sx, int sy, int
 }
 
 /**
- * @brief Scale up the rendered part of the back buffer to take up the full view
+ * @brief Scale up the top left part of the buffer 2x.
  */
 static void Zoom(CelOutputBuffer out)
 {
-	int wdt = gnScreenWidth / 2;
-
-	int src_x = gnScreenWidth / 2 - 1;
-	int dst_x = gnScreenWidth - 1;
-
+	int viewport_width = out.w();
+	int viewport_offset_x = 0;
 	if (PANELS_COVER) {
 		if (chrflag || questlog) {
-			wdt >>= 1;
-			src_x -= wdt;
+			viewport_width -= SPANEL_WIDTH;
+			viewport_offset_x = SPANEL_WIDTH;
 		} else if (invflag || sbookflag) {
-			wdt >>= 1;
-			src_x -= wdt;
-			dst_x -= SPANEL_WIDTH;
+			viewport_width -= SPANEL_WIDTH;
 		}
 	}
 
-	BYTE *src = out.at(src_x, gnViewportHeight / 2 - 1);
-	BYTE *dst = out.at(dst_x, gnViewportHeight - 1);
+	// We round to even for the source width and height.
+	// If the width / height was odd, we copy just one extra pixel / row later on.
+	const int src_width = (viewport_width + 1) / 2;
+	const int doubleable_width = viewport_width / 2;
+	const int src_height = (out.h() + 1) / 2;
+	const int doubleable_height = out.h() / 2;
+
+	BYTE *src = out.at(src_width - 1, src_height - 1);
+	BYTE *dst = out.at(viewport_offset_x + viewport_width - 1, out.h() - 1);
+	const bool odd_viewport_width = (viewport_width % 2) == 1;
 
-	for (int hgt = 0; hgt < gnViewportHeight / 2; hgt++) {
-		for (int i = 0; i < wdt; i++) {
+	for (int hgt = 0; hgt < doubleable_height; hgt++) {
+		// Double the pixels in the line.
+		for (int i = 0; i < doubleable_width; i++) {
 			*dst-- = *src;
 			*dst-- = *src;
-			src--;
+			--src;
+		}
+
+		// Copy a single extra pixel if the output width is odd.
+		if (odd_viewport_width) {
+			*dst-- = *src;
+			--src;
 		}
-		memcpy(dst - out.pitch(), dst, wdt * 2 + 1);
-		src -= out.pitch() - wdt;
-		dst -= 2 * (out.pitch() - wdt);
+
+		// Skip the rest of the source line.
+		src -= (out.pitch() - src_width);
+
+		// Double the line.
+		memcpy(dst - out.pitch() + 1, dst + 1, viewport_width);
+
+		// Skip the rest of the destination line.
+		dst -= 2 * out.pitch() - viewport_width;

```

Comment:
The number of comments and "+1", "-1" etc in this function makes it really scary and indicates to me that there must be a cleaner/more intuitive way to handle it.

Reply:
I hope so, you're welcome to give it a go :)

Reply:
Having poked it earlier, I take all the comments as a sign that this is deceptively non-simple :)

Reply:
I don't mean to imply it is "simple", just that we may be lacking more abstraction to solve it in a cleaner, more semantic way.

I don't have a suggestion was well.. it was just a gut feeling of "something is wrong here".